### ISSUE TITLE
WLED 0.15.0 Introduces Flashing/Noise on GPIOs 1, 16 on QuinLED-Dig-Quad

### ISSUE THREAD

**Issue by hacustom** [wled/WLED#4389]
### What happened?

After upgrading a QuinLED-Dig-Quad to WLED 0.15.0, LED segments on GPIOs 1 and 16 experience significant noise and flashing.  An LED segment on GPIO 3 operates fine without any issue.

Segment 1 GPIO 3 - 629 LEDs
Segment 2 GPIO 1 - 161 LEDs
Segment 3 GPIO 16 - 171 LEDs

This particular setup has been running flawlessly for just over 3 years.  The issue was immediately resolved by downgrading to WLED 0.14.3.

### To Reproduce Bug

- Upgrade a QuinLED-Dig-Quad to WLED 0.15.0
- Activate any preset or animation
- Flashing/Noise is prominent on LED segments on GPIO 1 and 16

### Expected Behavior

- No noise/flashing on any LED segment

### Install Method

Binary from WLED.me

### What version of WLED?

WLED 0.15.0 2412100

### Which microcontroller/board are you seeing the problem on?

ESP32

### Relevant log/trace output

```shell
I don't have any logs at the moment, I have since downgraded to WLED 0.14.3.  However, I would be more than happy to re-upgrade and capture any logs/traces/configs that would aid in troubleshooting.
```

### Anything else?

This appears to be similar to report #3976.  That was submitted for 0.15.0b2 and not the full release however.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by blazoncek** [comment]
GPIO 1 & 16 will use RMT. RMT is susceptible to flickering in some circumstances.
0.15 added a lot of code that may make those circumstances more frequent.

GPIO 3 uses I2S.

**Comment by hacustom** [comment]
Interesting.  

I had a spare Dig-Quad on hand, and I upgraded it to WLED 0.15.0 and configured the LED outputs to be identical as the main one for my house lights.  I wired up a spare string of LEDs I had leftover and I'm seeing the same behavior as with my other - tons of flashing/noise.  If I stick to a solid color, it's OK.  There's a bunch of noise every time I change color but once it reaches the target color the noise stops.  Any animation is unusable.

**Comment by hacustom** [comment]
Just for kicks, I setup my spare Dig-Quad from scratch, installing it from https://install.quinled.info/dig-quad/ just to see if it made any difference and it didn't.

**Comment by blazoncek** [comment]
If you want to test (and help with resolving) set up more than 4 outputs with less than 300 pixels each and report.

**Comment by eMeF1** [comment]
I encountered the same issue after upgrading to 0.15.0. There are 182 LEDs on GPIO 16, which are set to a solid orange-ish RGB[115,74,0] color. It's not apparent on solid white. I have a couple of dig2analog on GPIO 3, but I did not test whether the issue is also there. After downgrading to 0.14.3, the random flickering (every 30-60 seconds) issue is gone.

**Comment by blazoncek** [comment]
Periodic issue leads me to believe this may be network related.

**Comment by hacustom** [comment]
> If you want to test (and help with resolving) set up more than 4 outputs with less than 300 pixels each and report.

I will see if I can do that on the dig-quad as a second test.

**Comment by kylewilliams321** [comment]
Same issue from me and similar setup as original reporter

**Comment by intermittech** [comment]
If needed I can build a quick test setup and can troubleshoot on Discord too, apologies this made it through the beta somehow, I had not noticed in on any testing setups I had before. Let me know if you need anything!

**Comment by blazoncek** [comment]
Please test a set-up with at least 6 outputs (WS281x). You do not need to have actual LEDs connected or GPIOs attached. It is important to have GPIO 1, 3 & 16 as first 3 outputs and each output must not exceed 300 LEDs

**Comment by dokalma** [comment]
I had exactly the same problem on my digquad pushing 824 seedpixels per channel. 

Went back to 14.3 and all was good

**Comment by hacustom** [comment]
> If needed I can build a quick test setup and can troubleshoot on Discord too, apologies this made it through the beta somehow, I had not noticed in on any testing setups I had before. Let me know if you need anything!

Oh my goodness, PLEASE don't apologize!  I have worked in IT & Software Dev for...ouch...nearly 30 years now?!  I've been on both sides of bug reports more times than I can count, I actually have grown to love really gnarly bugs and digging into them.

Especially with you all that contribute to open source projects, the work you do is amazing and usually pretty damn thankless.  The only intent behind my bug report is to share information and see if we can figure out the cause :) 

I'm working on re-configuring my spare Dig-Quad right now to test out @blazoncek suggestion.

**Comment by hacustom** [comment]
OK, got my spare Dig-Quad setup like this:

6 LED Outputs, 190 LEDs on each output, WS281x, 30mA/12v.

Output 1 - GPIO 3
Output 2 - GPIO 16
Output 3 - GPIO 1
Output 4 - GPIO 4
Output 5 - GPIO 5
Output 6 - GPIO 13

I've been running a few different twinkle effects for ~10 minutes now without any apparent noise, using a single color and different color palettes just for testing.

**Comment by hacustom** [comment]
Just for kicks I deleted Output 6 and the noise was immediately re-introduced.

**Comment by Jean-Luc-Hauser** [comment]
Same issues with the Quad and 12V WS2815 on version 0.15 : heavy noises and flickering with effects such as Dancing Shadows.

GPIO 3 - 152 LEDs (60/m)
GPIO 16 - 356 LEDs (144/m)
GPIO 15 - Relay

I have a 5V PS for the relay and the Quad board and 12V for the LEDS through the board.

And I was not able to use the relay (GPIO 15) since it was taken by the I2S audio interface.

I have 2 GLEDOPTO GL-C-015WL-M controllers that have no issues with the 0.15 release. I went back to 0.14.3 for now, was scared to damage the LEDs

**Comment by blazoncek** [comment]
with 6 outputs or more, 1st 8 are using I2S (parallel)
with 5 or less, 1st is using I2S (single) and 4 are using RMT.
may notify @makuna if he has any clue what's different in NPB 2.8.0 compared to 2.7.4 in 0.14.4 regarding RMT

**Comment by travi** [comment]
I ran into what this sounds like with one of my digunos. I'm currently running 3 digunos and one digquad, but only saw this on one diguno and was able to fix it by swapping the esp from a spare diguno. 

A detail that I noticed was unique on that instance was that I had the output order switched around so that the first output was mapped in wled as gpio 3 and the second to gpio 16 instead of the other way around, for whatever reason. 

I imagine swapping the esp worked in my case only because that one had the outputs orders as gpio 16 then 3.

**Comment by promess** [comment]
I'm dealing with this same issue, except it's on a single ESP32. There are 260 LEDs in this configuration, all of which are in GPIO 16.

it's only the last section where I'm adding power back in after a split and it's after that point that the LEDs are misfiring. I've posted video of it flashing blue and red after a few minutes in the discord channel. It was working fine in 14.4. I'm not sure if I'm sending too much power and the additional power is fucking with it or something. Let me know if there's anything you need me to do.

**Comment by blazoncek** [comment]
FWIIW I'm using Dig-Quad with 16 (465), 1 (270), 3 (15) - in that order - with no flickering whatsoever.
Using generic ESP32 in Wemos D1 Mini size packaging from aliexpress.

**Comment by Makuna** [comment]
> with 6 outputs or more, 1st 8 are using I2S (parallel) with 5 or less, 1st is using I2S (single) and 4 are using RMT. may notify @Makuna if he has any clue what's different in NPB 2.8.0 compared to 2.7.4 in 0.14.4 regarding RMT

The only real change with RMT between those versions is the ability to compile using IDF5.  This was just some minor headers include changes to react to the IDF change but still outputs warnings about deprecated API use.
Did you turn on ESP32 core debug errors/warnings/info level to see if anything shows up in the debug output?

**Comment by JorinL** [comment]
Hey there,
I have the same problem on my 3 PWM strips.
When I cut the transitiontime it is gone - with transition it flashes when turning on before dimming up and when turning off while dimming down a flash and dimming further down.

I added a small video of the current situation.
https://github.com/user-attachments/assets/f6009672-1853-451c-b2fd-f404a0fe5acc

**Comment by blazoncek** [comment]
@JorinL PWM is unrelated to digital flickering. Please open a new issue.

**Comment by Woutch** [comment]
Hi All,

I have the same problem in my Chrismas tree. This is a Quinled Dig uno v3 like this:
https://shop.allnetchina.cn/products/quinled-dig-uno-v3r7-digital-led-controller

It feeds 2 strings starting from the middle and string one is the upper one and 2 is the lower one.
Using the latest v14.4 I have zero problem, upgrading to v15 gives the first 20 leds a noicelike color and keep flikkering, even in solid mode, but faster when an effect is activated. 

Output 1 is using gpio 1, and also tried 16, both same flickering.
Output 2 uses gpio 3, close to 0 problems. 

Limiting the brightness seems to help, but this is brightness at 10% and problems still are like 50% more ore less.

V14.4:
![20241215_194508](https://github.com/user-attachments/assets/8dae50e0-ea30-442b-8dac-c67b1e3854d5)

V15:
![20241217_004731](https://github.com/user-attachments/assets/0de4bb0c-e726-44a1-8c50-e630d091b376)

With effect:
https://github.com/user-attachments/assets/ef06e29d-556d-4a89-9d80-b0d910aa616f

Solid color:
https://github.com/user-attachments/assets/e6309d58-4a9c-48f4-b7d0-670ef472287c

If anyone needs more info, ask ;-)

Hope this helps,
Wouter

**Comment by Zappelphilipp** [comment]
same problem here. using 699 LED's on an AZDelivery ESP32 on GPIO 16. LED's are usually dimmed to about 10% on a solid color, after the update i get basically random color flashing everywhere. 

By dimming the Power above about 50% the random color flashing stops but there is still some flickering everywhere (at least in the right color)

sadly a downgrade seems not to work because after that the ESP seems to crash after a few minutes.

**Comment by hacustom** [comment]
Curious if anyone has had any luck with this?  I am still running an older version of WLED, not the biggest issue in the world, I am not exposing these externally, but I would like to keep current with them.

**Comment by miketunes** [comment]
I also have this issue, asked about it on the Quinled discord and they said to check here.

**Comment by Arcitec** [comment]
I was thinking of using ESP32-WROOM-32U pins 2, 4, 12, 16.

If I understand correctly, ESP32 has a library called "RMT" which is for controlling LEDs.

Is that library only being used with certain output pins, or certain combinations?

Unfortunately it's not fully clear to me if my pin choices will be affected by this situation.

**Comment by blazoncek** [comment]
ESP32 has a GPIO muxer. Any pin can be anything: PWM, I2S, I2C, RMT etc.
BTW RMT is not a library. It is a DMA HW driver.

**Comment by MakerMatrix** [comment]
Same issue here. Recent model dig-Quad running two strands of 12V BTF WS2814 (128 and 169 logical LEDs). LED data pins 1, 3, and 4 (GPIO 16, 1, and 4 respectively) all flicker badly on 0.15.0.  Moving the data pins around and swapping strips in and out shows that the issue is coming from the controller itself and not the LEDs or my wiring.

0.14.4 Audioreactive works very cleanly on all outputs.

Is this issue explained yet, if not fixed in 0.15.1 RCs?

**Comment by intermittech** [comment]
I believe it's fixed in v0.15.1-beta2, could you give that one a try @MakerMatrix ?

**Comment by blazoncek** [comment]
@MakerMatrix yes it was explained (elsewhere). GPIO don't matter but number of LEDs do. If you use more than 300 LEDs per output, WLED will use RMT peripheral for output instead of I2S (behaviour from 0.14).

**Comment by blazoncek** [comment]
No. That's for currently released 0.15.0, for 0.15.1 you'll need to check the changelog.
To overcome the issue on 0.15.0 set last output to contain more than 300 LEDs. This will force 0.14 behaviour.

**Comment by hacustom** [comment]
Ah, ok. I did misunderstand, I thought you meant that changing the pixels to be <300 per GPIO was going to be the only option in version 0.15.1 and beyond. Thank you for clarifying, much appreciated.

**Comment by MakerMatrix** [comment]
> [@MakerMatrix](https://github.com/MakerMatrix) yes it was explained (elsewhere). GPIO don't matter but number of LEDs do. If you use more than 300 LEDs per output, WLED will use RMT peripheral for output instead of I2S (behaviour from 0.14).

300 logical LEDs or 300 physical?  My longest strip has 169 pixels/507 leds.

I'll give -beta2 a try and report back.

**Comment by MakerMatrix** [comment]
I am sorry to report that 0.15.1-beta2 does not fix the problem I am hitting.  The "blends" effect makes the flickering most readily and immediately apparent.

**Comment by MakerMatrix** [comment]
OK I see it in the changelog for 0.16-alpha (the nightly build) so let's get that mentioned in here too.

WS281x leds flickering every firmware after WLED 0.15.0-b3 https://github.com/wled/WLED/issues/4211

**Comment by blazoncek** [comment]
Try different ESP (from another manufacturer).

**Comment by MakerMatrix** [comment]
> Try different ESP (from another manufacturer).

IMHO the dig-quad is the best hardware for the task at hand.  I believe it's important that WLED give users more control over when and if the new rendering peripheral is engaged since in my studies over the past few days it is giving a lot of people headaches, not only QuinLED customers (for example #4211 mentioned above).  I tried the suggested workaround to make the last LED entry >300 on 0.15.0 but either didn't understand that entirely, or it didn't work.

For now everything is happy on 0.14.4-Soundreactive but I'd like to not get orphaned beyond that release.  I say this as someone who has donated to the project, not just a total freeloader. ;-)

In any case the release notes indicate that it's fixed, somehow, in 0.16-alpha so until then...

**Comment by MakerMatrix** [comment]
@intermittech would it be possible to get a custom build of 0.15.x that disables WLED_USE_PARALLEL_I2S to put on your git?  I think there is a chance that would fix it?

https://wled.discourse.group/t/using-wled-use-parallel-i2s/12243

Edit: oof it looks like that #ifdef got removed at some point.

**Comment by willmmiles** [comment]
I'm able to replicate this on my test system, even with the latest 0.16 build.  There seems to be an issue with the Parallel I2S hardware driver which results in occasional output timing errors during web server accesses.  I am investigating.

**Comment by intermittech** [comment]
Interesting, from my understanding the main issue was fixed in v0.15.1-beta2 so I'm not 100% sure what's going on, I will give it a try sometime soon (currently in a crunch to get a video out)!

**Comment by MakerMatrix** [comment]
> Interesting, from my understanding the main issue was fixed in v0.15.1-beta2 so I'm not 100% sure what's going on, I will give it a try sometime soon (currently in a crunch to get a video out)!

After spending hours on this, I'm not even sure if I'm hitting the Parallel I2S issue, TBH.  I've tried the workaround of creating a fake strip with >300LEDs, which by all accounts *should* result in disablement of that
https://github.com/wled/WLED/blob/23a51e0982a5aaab02371246fc38e11863dcde5d/wled00/FX_fcn.cpp#L1149-L1150
But it changes nothing.  Then there is the fact that GPIO3 works fine in all conditions, and the other three flicker in all conditions.  Very bizarre.  I've also documented a lot of this in the troubleshooting discord (Jarrod).

**Comment by blazoncek** [comment]
@willmmiles according to @Makuna experiment with cadence 4 for parallel I2S prior to do any other investigation. Cadence 3 and cadence 4 have slightly different timings which may behave differently on edge cases (out of specs systems, plenty of IRQs, etc). Changing cadence requires compile override.

@MakerMatrix change to different hardware (controller board and ESP) to rule out HW/SW issue.

Out of my 10+ production units, none exhibits any visual glitch. One or many outputs, <100 LEDs and >500 LEDs per output.
I use variety of HW form @intermittech, @srg74, AthomTech and a few other OS boards, including some soldered at home.
My WiFi network contains over 100 units, constantly chatting.

**Comment by willmmiles** [comment]
@blazoncek Right, let me walk it all back.  The problem I'm seeing is with the RMT driver -- I had the test inverted in my debug code.  Parallel I2S is rock solid by comparison. :)   (I do think there is a possible weakness to interrupt latency in NPB's I2S DMA logic; but this doesn't seem to be a problem in practice.)

The glitches I am seeing are induced by network traffic, and look to me like a FIFO underrun: at some point partway through the output, the data "slips" one (or more) bit times, resulting in shifted or corrupted colors.  Tellingly, the glitch events can cause LEDs past the end of the configured count to light.

I've reproduced this on a couple of different ESP32 (WROOM/WROVER) boards from different manufacturers.

I'm digging in to the RMT driver now.

**Comment by MakerMatrix** [comment]
> I can try to upgrade my main house lights Dig-Quad to the beta this evening to see if it still exhibits the behavior for me. I had a test instance setup on a bare ESP32 that I was able to replicate the behavior on too I think.

I'm curious.  You and I seem to have the *exact* same issue in that GPIO3 works fine, the others not.  Did you test GPIO 4?

So: ESP HW experts - what's different about GPIO3, in hardware, compared to the other three that @intermittech exposes on the dig-Quad as LED1-LED4 data pins?  Maybe there is a clue hiding there...

https://quinled.info/quinled-dig-quad-pinout-guide/

Edit: I see that @blazoncek answered that in the first reply (they use RMT).  Doh.  I am just trying to get my head around why, when I create a third "virtual" strand of lights with >300, I still have flickering on the other pins while GPIO3 is stable.

Edit2: RMT != Parallel I2S.  I'm not sure how or why I conflated those....learning though.

**Comment by willmmiles** [comment]
The first defined LED output will use the I2S driver (ie. the one at the top of the LED configuration page).  Subsequent outputs will use the RMT driver instead.  If you remap the pins between the different outputs, I would expect that you see the first one is consistently stable, regardless of what pin it is assigned to, and the others glitchy.

**Comment by MakerMatrix** [comment]
That might be a revalation.  If I had not seen this with my own eyes and spent 20 hours on troubleshooting, I'd not believe this.  But here is the behavior on 0.15.0:

First LED output (169LEDs) = GPIO3
Second LED ouput (169LEDs) = GPIO16
GPIO16 flickers

First LED ouput = GPIO16
Second LED output = GPIO3
Neither one flickers!!

The moment I hit "Save" after switching them back to the top config in the UI, flickering on GPIO16 begins.

@hacustom do you see this?

**Comment by phoenix1589** [comment]
I've got similar symptoms, which I had put down to a poor quality setup (it's not a QuinLED board, I don't have level shifters etc) however the recent comments especially around the 300 LED mark have led me to do a couple more tests.

Running WLED 0.15.0 (build 2412100) on a LILYGO ESP32-Wrover, with LEDs setup as:
219 LEDs on GPIO 26
130 LEDs on GPIO 3
33 LEDs on GPIO 23

with then 2 segments of 219 LEDs and 163 LEDs (i.e. output 1 as 1 segments and output 2+3 as the other segment)

I would get frequent flashes every few seconds. 

If I change the 2nd segment down to 80 LEDs (bringing the total to 299) then the flashing stops. If I then set the 2nd segment back to 163 - the flashing does not return. If I then change the effect, or turn off back on the outputs - the flashing returns.

Doing the change again, the flashing stops again.

**Comment by blazoncek** [comment]
Nice chat (revealing). TL:DR you may want to force parallel I2S which is triggered by using more than 5 outputs.
Remember, they need to be less than 300 pixels in length.

When you have >5 outputs WLED will switch to parallel I2S for firs 8 outputs and use RMT for last 8 (if configured). Unfortunately parallel I2S uses a lot of buffers so we needed to enforce a limit of 300 pixels to prevent memory drain.

As far as GPIO go, all IO are equal (ESP32 uses GPIO muxer).

**Comment by MakerMatrix** [comment]
I still don't completely understand (Is there a primer I can read about these different drivers?) but feel like we are making progress because....

I added four dummy outputs of 10 each, and then removed these 40 fake pixels from segment 0.  Then I swapped the GPIOs back to the configuration where I previously had flicker and...drumroll...NO FLICKER.

**Comment by MakerMatrix** [comment]
> The glitches I am seeing are induced by network traffic, and look to me like a FIFO underrun: at some point partway through the output, the data "slips" one (or more) bit times, resulting in shifted or corrupted colors. Tellingly, the glitch events can cause LEDs past the end of the configured count to light.

Ok, this is very interesting and telling.  During initial setup of this new rig, I was on 0.15.0, and I had a bit of a time figuring out what was the correct number of LED's on my edited GPIO16 strip (I wasn't going to count 169 of them by hand).  First I thought it was 165.  I'm now convinced this was because the last four had erroneously lit up when I crept up on the end.  Then, after a couple reboots later, the last few lights were not switching on and I had to add more, which really had me puzzled.  It might have been that at some point in between I had switched to 0.14.4 once I noticed flickering.  I had put all of that out of my mind before reading this paragraph of yours a second time.

**Comment by Makuna** [comment]
> If I change the 2nd segment down to 80 LEDs (bringing the total to 299) then the flashing stops. If I then set the 2nd segment back to 163 - the flashing does not return. If I then change the effect, or turn off back on the outputs - the flashing returns.
> 

Is the flashing only on LEDs "after" the 80th or on all LEDs on the strip?

**Comment by hacustom** [comment]
> That might be a revalation. If I had not seen this with my own eyes and spent 20 hours on troubleshooting, I'd not believe this. But here is the behavior on 0.15.0:
> 
> First LED output (169LEDs) = GPIO3 Second LED ouput (169LEDs) = GPIO16 GPIO16 flickers
> 
> First LED ouput = GPIO16 Second LED output = GPIO3 Neither one flickers!!
> 
> The moment I hit "Save" after switching them back to the top config in the UI, flickering on GPIO16 begins.
> 
> [@hacustom](https://github.com/hacustom) do you see this?

@MakerMatrix work was busier than I had anticipated, I didn’t have time to do an upgrade yet. I know I had tested a few scenarios out when I first ran into the issue but I don’t remember the specifics anymore. I think I documented some of it in my earlier comments. I only am looking on my phone right now so I might be missing it.

**Comment by hacustom** [comment]
Installed WLED_0.15.1.beta2_ESP32.bin on my main Dig-Quad.  Flashing/flickering is still present on GPIO's 1 and 16.  

Segment 1 GPIO 3 - 629 LEDs
Segment 2 GPIO 1 - 161 LEDs
Segment 3 GPIO 16 - 171 LEDs

Just downgraded back to 0.14.3 and all continues to be well there.

**Comment by blazoncek** [comment]
@hacustom why are you regressing? You already [reported](https://github.com/wled/WLED/issues/4389#issuecomment-2543306676) that 6+ outputs resolves the issue.

It is now apparent that the root cause is RMT driver (and not I2S as I thought on a first glance after a revisit).

@willmmiles I suggest you don't spend a lot of time debugging RMT driver as it is far easier to (always) use parallel I2S. Especially once @Makuna manages to reduce backend buffers (which will allow larger LED count without depleting RAM).

@Makuna I've implemented complete pixel buffer management in/for 0.16 (to allow layer blending) and we no longer rely on NPB to provide `GetPixelColor()` or maintain double buffers. So it might be possible to instruct NPB to use single buffer (either during creation or `begin()` or even at compile time) if you ever want to explore that (I have not looked at new additions to NPB in quite some time so please excuse me if that is already implemented).

For anyone experiencing this issue with any version of 0.15, please revert to 0.14 and wait for 0.16 or create additional phantom outputs (1 pixel is enough) while keeping other outputs with less than 300 pixels. I don't think we'll backport RMT solution into 0.15 as it is already waiting in beta for quite some time without any proper feedback from beta testers and 0.16 will bring so many enhancements and performance improvements that it makes no sense to invest time in 0.15.x.

**Comment by phoenix1589** [comment]
> > If I change the 2nd segment down to 80 LEDs (bringing the total to 299) then the flashing stops. If I then set the 2nd segment back to 163 - the flashing does not return. If I then change the effect, or turn off back on the outputs - the flashing returns.
> 
> Is the flashing only on LEDs "after" the 80th or on all LEDs on the strip?

The flashing has always occurred on all LED strips. Neither strip individually has more than 300 and this testing was changing just the segment configuration of LED count.

I have now created 3 additional LED Outputs (total of 6) with the 3 additional each have 1 LED and a random GPIO. Having the full 382 LEDs configured there is consistently no flashing.

**Comment by hacustom** [comment]
If adding phantom segments is the full and final fix, that works for me and we can close this issue out. I will just not be staying on 0.14.3 and isolating these devices so that only local control is possible.

**Comment by willmmiles** [comment]
I'm not satisifed with "just don't use RMT"; I'd like to make sure I fully understand the depth and scope of the issue, and that's not a solution for anyone with longer output lengths.   This used to work on 0.14 which was not so long ago - what did we change that triggered this regression?

I've been doing a survey of "what do other projects do"; it seems like our friends over at MoonModules had, at some point, observed some trouble with I2S -- enough that they've got a compile-time flag to set the default assignment to RMT first.  It looks like FastLED actually implements "detect and abort" logic if there's an RMT interrupt timing issue.  And of course there's an entirely new RMT driver scheme provided by IDF v5 to support the improved capabilities of new chips -- it looks like Espressif has added DMA hardware for the RMTs in their newer chip designs expressly to address problems with interrupt latency for our kind of use case.

As for backporting to 0.15.1: I believe that this is a big enough problem to warrant a real fix, even if we're just changing the default hardware assignment to parallel I2S as a mitigation.  Whatever it turns out to be, I don't expect it to be a difficult port, and I think we'd be doing users a big disservice if we ship something this broken as the next release.

Long at the short of it: I'm still investigating and I'll hopefully have more answers as to the root cause and possible mitigation soon.

**Comment by MakerMatrix** [comment]
I would tend to agree - or as I said above, at least provide some documentation and a clean way to see and control what bus gets used for which outputs.  The hacky and unclear workaround of a known bug shouldn't be the only option in production code.  I lost many hours to it.

A good start would be to report which outputs are programmed and what kind of bus they use, under (i)->Instances for each instance.  That might be a great boon to users and devs alike for troubleshooting issues.

**Comment by Makuna** [comment]
At a low level, NeoPixelBus (my library) is used to drive the LEDs.  RMT implementation uses exposed APIs that on one of the last core updates got depreciated and further broke it as it gets interrupted by other ISRs (like WiFi).  NPB got updated to the new APIs, which still had issues.  The fix for those came in a very new core update.  But my understanding is that this newer core uses more memory making WLED not feasible on it at the moment (@blazoncek will need to confirm this).

**Comment by willmmiles** [comment]
@Makuna Do you know offhand of any particular platform cores that are known to be problematic (or known to be fixed)?  My failing builds are using tasmota 2023.06.02 (IDF 4.4.4).  I'm happy to test newer cores - I'm confident we can find solutions to mitigate space issues.

**Comment by blazoncek** [comment]
@MakerMatrix if the port would be simple I would agree with you, but it is not (at least the way I see it ATM). However you are welcome to join the efforts.

@willmmiles we should be a few (couple of?) months away from releasing 0.16 so IMO we should focus on that and try to expedite it instead. Not everyone is hitting RMT bug (in fact only users with 2 to 5 outputs may be effected by it and not all of them will see it). But I understand the desire to *hunt down the nasty bug*.

@Makuna my parallel I2S tests were performed in April/May 2024 (NPB versions 2.7.9 and 2.8.0). I am currently using 2.8.3 in my development fork but did not do any memory testing on it. IIRC the memory calculation formula was `channels * max_pixels_per_channel * 8 * 3` where 8 comes from parallel I2S and 3 comes from three buffers (front/back/transmit) but I can't find my notes any more. However, if memory serves me well, if I used more than 300 pixels memory on an ESP32 got so low that it would crash due to OOM exception during WiFi activity which doubles the above formula. I know you were working on some modifications to backend (coreshader?) but didn't follow due to the lack of time.
Looking back at commit history [I favoured I2S](https://github.com/wled/WLED/commit/20ed81cd86a7eb3ee1ffcfad4d7c3b2cf84d2a3a) after the switch from [2.7.5](https://github.com/wled/WLED/commit/0aea75edb78febff455afb82589f84ad188fb81f) to [2.7.8](https://github.com/wled/WLED/commit/c74db95c1430e2998eae99996c01f8b2de5dca9a) and this might have been due to user reports of [flashing](https://github.com/wled/WLED/blob/20ed81cd86a7eb3ee1ffcfad4d7c3b2cf84d2a3a/wled00/bus_wrapper.h#L1498) glitches. 

@hacustom I was explaining a workaround anyone stuck with (or needing) 0.15 may want to try. There is no need to modify "segments" as you wrongly explained (but deleted the post). If you are selling products to your customers that use WLED then you should fork and only use your fork's binaries and keep that locked. Or invest time and help develop a correct solution.

**Comment by jasondostal** [comment]
Appreciate your thoughts on that. If I do ever make the decision to sell services or products utilizing WLED, it’s certainly good advice. I tend to follow the same practice when a component, package, library or module becomes critical at work. 

For projects I work with on a personal basis I tend to contribute via donations.

**Comment by jasondostal** [comment]
Are you aware of any commercial hardware manufacturers that utilize WLED that follow such a practice?  I currently use a mix of ESP32s and Dig-Quads but if there’s an alternative product that takes an approach like that, I wouldn’t be opposed to trying them out.

**Comment by MakerMatrix** [comment]
I did not notice this until just now because I was focused on flicker, but it appears that with the workaround, my 12V 60/m BTS WS2814 does not render blue, at all.  I guess I'm done with 0.15.x

**Comment by blazoncek** [comment]
This has been brought up in the past. @Makuna any thoughts why would RGBW not work correctly with parallel I2S?

**Comment by willmmiles** [comment]
A quick update:
- I tried several schemes to detect and recover within the context of the ESP-IDF RMT driver currently in use by NeoPixelBus.  Unfortunately, none of them were satisfactory: once the RMT driver is out of step, there's no possibilty to recover it, and round-tripping recovery through another task (even at a high priority!!) takes too long to avoid visible flickering.
- I have started prototyping a replacement RMT driver to implement fast recovery within the ISR, but I have yet to solve all problems with it.
- Another option may be to leverage the memory of unused RMT devices -- eg. if we need only 2 outputs, they can each use 4x the available RMT memory, reducing the risk of underflow.  This would require a more sophisticated resource allocation strategy from WLED, though, and wouldn't help much with some systems (S2s and C3s have only 4 RMTs to work with).

**Comment by blazoncek** [comment]
One thing that popped during today's teleconference: excessive IRAM_ATTR used in functions for pixel drawing.
One option would be to remove IRAM_ATTR from every (non ISR) function of WLED and see what that brings.
Use custom compile with `-D WLED_SAVE_IRAM` compile flag.

**Comment by willmmiles** [comment]
Unfortunately, I found adjusting the IRAM allocation had no effect.  :(

It's been a bit of an adventure to get good traces of what's going on -- ESP-IDF v4 doesn't support post-mortem system tracing, and real-time tracing disturbs the timing too much; I've kind of had to hack it a bit.  Sadly, the results are not encouraging: I'm seeing the FreeRTOS task scheduler exceeding the latency required for RMTs with WS2812s.   I think perhaps what's happened is that changes to the platform have increased the task scheduler latency.   It's possible this can be tightened up some, but it's at best probabilistic, and not a cure.  A quick eyeball check of IDF v5 suggests there aren't any obvious changes to the scheduler code in newer versions, but it's hard to be sure.  I've also seen some hints that there are some chip performance settings during bootup that have been changed (for better or for worse) over the version history that might have an impact here.

I don't think there's any path forward for reliable RMT output short of boosting the interrupt priority above FreeRTOS; this is technically possible, but to paraphrase the ESP manual, "some assembly required".  I'm going to sketch that next.  I'm not confident such a solution will win upstream NeoPixelBus approval, but I'm in it for the experience at this point ;)

**Comment by Makuna** [comment]
Just a little technical background on RMT on ESP platforms.  There is a limited amount of DMA memory dedicated to the RMT usage; so little it is useless for WS2812x driving directly using true ISRs.  So, what is relied on is a "pumping" system where, as the DMA buffer is emptied, it gets updated/loaded from RAM (IRAM was better for latency).  The only part that NPB implements is the translation and handoff of the data (using IRAM) for the system to then insert it into the DMA memory buffers they manage.  Too much latency in this translation and you will get glitches.  Since this pumping is done as a service (again, not implemented by NPB) and not a real ISR, it is sensitive to resource starvation.

This "model" (API and how it is done underneath) changed from one core version to the newer ones.  Alot of problems got introduced by that change.  So, knowing which "model" is being used is important.

**Comment by willmmiles** [comment]
I take it by "DMA" you mean the limited amount of RMT output buffer memory.  I fully agree - the RMTs really are just one step up from bit-banging with those tiny little buffers that need to be serviced every 38us or so.  It's my understanding that in some newer chips (ESP32-S3 and P4), some general-purpose DMA hardware ("GDMA") was added that can be programmed to refill the RMT buffer automatically, alleviating this problem.  I believe that's one of the factors that prompted the RMT driver redesign in IDF v5.

Ultimately, both of Espressif's RMT drivers (old and new) -- and FastLED's custom driver, for that matter -- rely on the "low-priority" interrupts, which, alas, are vulnerable to latency from the FreeRTOS scheduler.  I'm hoping that a custom driver with a higher priority interrupt will be able to improve the reliability for the non-GDMA capable platforms, though at the cost of extra (assembly) code and complexity.  It'll also be interesting to see how it performs across the family; most of them have even *smaller* RMT buffers than the baseline ESP32, and of course the RISC-based ones may need a completely different solution (or not at all - could be FreeRTOS is just faster there).

**Comment by willmmiles** [comment]
I'm happy to report that the high-priority interrupt driver is working well on my first test board - I can now get stable output with any amount of network traffic.  I'm working on making sure it builds and operates on more chips.  With any luck I should be able to pass along a beta on the weekend.

**Comment by DedeHai** [comment]
that is great news

**Comment by netmindz** [comment]
Sounds like some good progress, but it's feeling like this is going to be a feature for 0.16 rather than a bugfix for 0.15.1

What is the recommendation to prevent/discourage WLED config that are going to deliver problematic user experience?

**Comment by blazoncek** [comment]
AFAIK the patch will require NPB update.

However I briefly tested parallel I2S with 600 pixels (with NPB 2.8.3 in my fork) and it indeed seemed the ESP is not starved of RAM. I do have the code finished in my dev branch including UI improvements. Unfortunately it may need a bit of testing which I'm not able to do ATM.

Anyone interested in helping/testing forced parallel I2S is welcome to pull from my dev branch and compile. Once you have ESP running, configure 2 or more outputs with the same type of LEDs and select/check "Use parallel I2S" option (checkbox). It is still based off of 0.15 codebase with a few modifications (I would not recommend using it in production systems).

**Comment by willmmiles** [comment]
> Sounds like some good progress, but it's feeling like this is going to be a feature for 0.16 rather than a bugfix for 0.15.1
>
> What is the recommendation to prevent/discourage WLED config that are going to deliver problematic user experience?

I haven't got a thorough analysis on all platforms - still building out that hardware test matrix.

What I know is:
- On a baseline ESP32, using IDF v4, the FreeRTOS scheduler can introduce enough latency with interrupts disabled to glitch an RMT output.  This could happen with any network I/O.  I don't have a full list of what other subsystems may create FreeRTOS tasks.
- Longer strips and more outputs are more vulnerable, as they take longer to output.  

Testing is still pending on:
- ESP32 with IDF v3.  I didn't see any significant code differences in the FreeRTOS integration, so it's likely also vulnerable.  From the commentary here I think it's safe to say it doesn't work reliably there either.
- ESP32-S2
- ESP32-S3
- ESP32-C3

I should have more results over the weekend - most of the boards have arrived, I still need to assemble the test jig.

Known safe configurations:
- Single strip output using I2S
- Parallel I2S; requires all strips to have the same timing, this is not validated by WLED, see #4315.  Also note that 0_15_x and main [currently choose to use all available RMT channels before using any parallel I2S, even when enabled](https://github.com/wled/WLED/blob/a25fc6e0980da00184cd142fbda3c61d1a3cb199/wled00/bus_wrapper.h#L1452).

Or to sum up: don't use RMT.  (This is, of course, no help to anyone who's got two strips with different timings connected to one controller.)

Sorry I don't have full information yet - I've been focusing more on a fix than building the test matrix.

**Comment by DedeHai** [comment]
Thanks for the summary.
I think a sporadic flicker on heavy traffic is accepted by users, otherwise there would be way more complaints. Also C3 glitches all the time when accessing the web UI, very few complain about that. My point being: it does not need to be perfect in 0.15, just not really bad with constant or very frequent flickering.

**Comment by blazoncek** [comment]
> * Parallel I2S; requires all strips to have the same timing, this is not validated by WLED,

I have a fix for that ready.

**Comment by willmmiles** [comment]
First test builds (based on 0.15.x branch) are up: https://github.com/willmmiles/WLED/actions/runs/16158643608
Use the "rmt" bins; I don't know why the CI system also zipped up older files along with them.  This should be considered quite experimental; please don't install them on a device that you can't easily reflash via USB.

So far I've tested ESP32 (IDF v3 and v4), and ESP32-S2.  Good chance S3 will also work.  C3 is not supported yet.  I've now received an S3 and C3 board for my bench but I don't have them set up yet.

**Comment by kylewilliams321** [comment]
I see this was addressed in 15.1 release so thanks everyone!

However, note mentions led segments need to be under 600 LEDs. My longest is 766. Should I wait or do I need to split this segment in 2?  (Not sure I can but probably could break off 100-150 to new segment possibly as it's permanent lighting up on my house and the controller is in garage on one side of house so the whole longest distance to first LED comes into play..)

**Comment by blazoncek** [comment]
600 is a limit that will definitely work for everyone.
You can, however change it and experiment.

**Comment by DedeHai** [comment]
@kylewilliams321 I would hold off with changing hardware if you are not in a hurry, we are working on it and it will eventually be fixed.

**Comment by MakerMatrix** [comment]
Apologies for disappearing...I intend to test the new RMT driver as soon as I get time to cycle back to my LED project.  It looks like this code made the 15.1 release and I no longer need the @willmmiles test builds?

Thank you for all the work that went into this!

**Comment by willmmiles** [comment]
> Apologies for disappearing...I intend to test the new RMT driver as soon as I get time to cycle back to my LED project. It looks like this code made the 15.1 release and I no longer need the [@willmmiles](https://github.com/willmmiles) test builds?
> 
> Thank you for all the work that went into this!

0.15.1 includes a mitigation (use parallel I2S instead of RMT) that works for <600 LEDs per bus and only if all busses have the same LED type.  This should fix the flickering in many configurations, but doesn't solve the problem in all cases.

Current status is that principal development on the new RMT driver is done for all hardware; it's ready for wider scale testing, code cleanup, and so forth.

**Comment by kylewilliams321** [comment]
Thanks team for all you do and really appreciate all the hard work being
done!

**Comment by MakerMatrix** [comment]
> Current status is that principal development on the new RMT driver is done for all hardware; it's ready for wider scale testing, code cleanup, and so forth.

I take that to mean I need your builds (I had already downloaded the right one for my setup) to test the new driver, then.  I'll try to look at this over the weekend.

**Comment by intermittech** [comment]
Just to chime in, I've been testing a build provided by @willmmiles that has the RMT bug fix on top of WLED v0.15.1 and that's been working excellent for me on a QuinLED Dig-Quad! So there are 2 fixes available then, using Parallel I2S (implemented in current WLED v0.15.1 but has some limitations) and "fixed RMT" which works well for me and allows mixing of LED types over various channels and such again (I setup a test with 2x 300 ws2812b, 1x 300 sk6812 and 1x ws2812 500 LEDs) and all flickering is gone (in any GPIO order)!

So a fix is close to this issue and then everyone can upgrade to v0.15.1 (or v0.15.2 if it includes the RMT fix also) without worry!

**Comment by kylewilliams321** [comment]
Great news!

**Comment by intermittech** [comment]
For people who want to test the RMT High Priority Interrupt Driver, it's integrated into the recently released WLED v0.15.2-beta1! The WLED github doesn't have pre-compiled binaries right now it seems but [we have made some for the QuinLED](https://github.com/intermittech/QuinLED-Firmware/tree/main) boards (or generic, the binaries are just pre-configured) if you want to give it a try!

**Comment by sintei** [comment]
> For people who want to test the RMT High Priority Interrupt Driver, it's integrated into the recently released WLED v0.15.2-beta1! The WLED github doesn't have pre-compiled binaries right now it seems but [we have made some for the QuinLED](https://github.com/intermittech/QuinLED-Firmware/tree/main) boards (or generic, the binaries are just pre-configured) if you want to give it a try!


I went from 13 to latest beta release and immediately had the flickering. Checked everything before googling and arriving here. I also tried precompiled binaries from you, but it's the same thing with the flickering.
Note, I also tried 33ohms switches.

**Comment by intermittech** [comment]
> > For people who want to test the RMT High Priority Interrupt Driver, it's integrated into the recently released WLED v0.15.2-beta1! The WLED github doesn't have pre-compiled binaries right now it seems but [we have made some for the QuinLED](https://github.com/intermittech/QuinLED-Firmware/tree/main) boards (or generic, the binaries are just pre-configured) if you want to give it a try!
> 
> I went from 13 to latest beta release and immediately had the flickering. Checked everything before googling and arriving here. I also tried precompiled binaries from you, but it's the same thing with the flickering. Note, I also tried 33ohms switches.

Have you tried with I2S parallel too?

**Comment by sintei** [comment]
> > > For people who want to test the RMT High Priority Interrupt Driver, it's integrated into the recently released WLED v0.15.2-beta1! The WLED github doesn't have pre-compiled binaries right now it seems but [we have made some for the QuinLED](https://github.com/intermittech/QuinLED-Firmware/tree/main) boards (or generic, the binaries are just pre-configured) if you want to give it a try!
> > 
> > 
> > I went from 13 to latest beta release and immediately had the flickering. Checked everything before googling and arriving here. I also tried precompiled binaries from you, but it's the same thing with the flickering. Note, I also tried 33ohms switches.
> 
> Have you tried with I2S parallel too?

I use pin 16, 3 and 4 in that order.
Each pin consists of 384 leds.
And I use the audiomicrophone as well.
The dig-quad is behind a barrier and for me to get in there to change physical pins is not something I'd rather do unless this is the only solution.
As of now, 14.3 works fine so I'll keep it until issue is resolved by all you smart people :)

**Comment by ywwg** [comment]
I reassigned the I2S pins according to this reddit comment and it worked: https://www.reddit.com/r/WLED/comments/1hbsc2u/comment/m1r64cz/

 I2S SD to port 17, I2S WS to ports 32 and I2S SCK to port 33

**Comment by intermittech** [comment]
> > > > For people who want to test the RMT High Priority Interrupt Driver, it's integrated into the recently released WLED v0.15.2-beta1! The WLED github doesn't have pre-compiled binaries right now it seems but [we have made some for the QuinLED](https://github.com/intermittech/QuinLED-Firmware/tree/main) boards (or generic, the binaries are just pre-configured) if you want to give it a try!
> > > 
> > > 
> > > I went from 13 to latest beta release and immediately had the flickering. Checked everything before googling and arriving here. I also tried precompiled binaries from you, but it's the same thing with the flickering. Note, I also tried 33ohms switches.
> > 
> > 
> > Have you tried with I2S parallel too?
> 
> I use pin 16, 3 and 4 in that order. Each pin consists of 384 leds. And I use the audiomicrophone as well. The dig-quad is behind a barrier and for me to get in there to change physical pins is not something I'd rather do unless this is the only solution. As of now, 14.3 works fine so I'll keep it until issue is resolved by all you smart people :)

Actually something you could try is changing that order, make GPIO16 not the first in the list, it would be interesting to see what that changes, if you are up for a bit of experimenting!

**Comment by sintei** [comment]
> > > > > For people who want to test the RMT High Priority Interrupt Driver, it's integrated into the recently released WLED v0.15.2-beta1! The WLED github doesn't have pre-compiled binaries right now it seems but [we have made some for the QuinLED](https://github.com/intermittech/QuinLED-Firmware/tree/main) boards (or generic, the binaries are just pre-configured) if you want to give it a try!
> > > > 
> > > > 
> > > > I went from 13 to latest beta release and immediately had the flickering. Checked everything before googling and arriving here. I also tried precompiled binaries from you, but it's the same thing with the flickering. Note, I also tried 33ohms switches.
> > > 
> > > 
> > > Have you tried with I2S parallel too?
> > 
> > 
> > I use pin 16, 3 and 4 in that order. Each pin consists of 384 leds. And I use the audiomicrophone as well. The dig-quad is behind a barrier and for me to get in there to change physical pins is not something I'd rather do unless this is the only solution. As of now, 14.3 works fine so I'll keep it until issue is resolved by all you smart people :)
> 
> Actually something you could try is changing that order, make GPIO16 not the first in the list, it would be interesting to see what that changes, if you are up for a bit of experimenting!

That's the least thing I can do!
I changed and set the order with 16 being in second or third "place". Unfortunately still issues.
Downgraded again.

**Comment by intermittech** [comment]
> > > > > > For people who want to test the RMT High Priority Interrupt Driver, it's integrated into the recently released WLED v0.15.2-beta1! The WLED github doesn't have pre-compiled binaries right now it seems but [we have made some for the QuinLED](https://github.com/intermittech/QuinLED-Firmware/tree/main) boards (or generic, the binaries are just pre-configured) if you want to give it a try!
> > > > > 
> > > > > 
> > > > > I went from 13 to latest beta release and immediately had the flickering. Checked everything before googling and arriving here. I also tried precompiled binaries from you, but it's the same thing with the flickering. Note, I also tried 33ohms switches.
> > > > 
> > > > 
> > > > Have you tried with I2S parallel too?
> > > 
> > > 
> > > I use pin 16, 3 and 4 in that order. Each pin consists of 384 leds. And I use the audiomicrophone as well. The dig-quad is behind a barrier and for me to get in there to change physical pins is not something I'd rather do unless this is the only solution. As of now, 14.3 works fine so I'll keep it until issue is resolved by all you smart people :)
> > 
> > 
> > Actually something you could try is changing that order, make GPIO16 not the first in the list, it would be interesting to see what that changes, if you are up for a bit of experimenting!
> 
> That's the least thing I can do! I changed and set the order with 16 being in second or third "place". Unfortunately still issues. Downgraded again.

Interesting, this seems to be a different kind of bug/issue then since before that would resolve or at least move the issue. Hmm...

**Comment by Jivemofo** [comment]
Still experiencing this issue on 0.15.3 with a specific configuration:
Hardware: QuinLED Dig-Uno (ESP32) with WS2812B 100-LED fairy lights
Configuration that causes flicker:

Single output on GPIO 16
Integrated with Home Assistant (WLED integration) and LedFX (UDP)
Flicker appears as random multicolored pixels, even on solid color presets

What I've ruled out:

Not hardware — tested multiple power supplies, re-wired connections
Not the GPIO pin itself — same flicker follows when switching to GPIO 3 as sole output
Not the presets JSON — deleted all presets, flicker persists

Workaround that works:

Adding a dummy output as output 1 (GPIO 16: Length 1, Skip 1) eliminates the flicker entirely on output 2.
With two outputs configured, the actual LEDs on GPIO 3 work perfectly

Key observation:
The flicker only occurs when using a single output configuration (and possibly combined with external integrations (HA polling + LedFX UDP streaming)). Adding any second output — even a dummy one — resolves it completely.
This suggests the RMT/I2S driver selection or buffer handling behaves differently with single vs. multiple outputs, and the single-output path still has issues under network load.

This comment was drafted with assistance from Claude AI, with human oversight and review.

**Comment by DedeHai** [comment]
@Jivemofo this is not related to this issue, the first output uses I2S which is not influenced by network traffic. What this does show however is the core of most if not all reported flickering issues with 0.15.3 and 0.16: the timings of I2S and RMT are not identical. While both are well within specification of most LEDs, adding cables and a whole bag of LED revisions, manufacturers and production tolerances can make I2S or RMT a better timing choice, fully depending on the specific setup. This is an unsolvable issue without adding more bus types with different timing options. What those options should be is still not entirely clear to me as there are just too many contradicting reports. I may just take a stab at it at random.

**Comment by Jivemofo** [comment]
@DedeHai ahh, that makes a lot of sense, thank you for your insight. I was really scratching my head on what was going on.
